### DISCUSSION TITLE
INET 4 and IEEE 802.1AS (gPTP)

### DISCUSSION THREAD

**Discussion by HpLightcorner** [inet-framework/inet#674] [General]
Hi INET Team,

I am currently investigating Omnet functionality for my PhD work and INET seems to be a perfect option to simulate my proposed network, which shall use several features of the IEEE TSN standards. In addition, I will take a deep dive into advantages of hardware (L1) syntonized clocks - like proposed by the White Rabbit Network or IEEE 1588-2019 (but in a reduced implementation).

Hence, especially the clock modules available with the latest release are of interest. I am currently planning to start with implementing the IEEE 802.1AS (gPTP) protocol, and this [paper ](https://easychair.org/publications/open/Q4kL) gives an excellent starting point. In their outlook, the authors are proposing to use the message dispatcher functionality.

INET is still quite new to me and I am fiddling out how everything is working, but before i start with the implementation I would like to now if anyone is already focusing on TSN protocols (what is done, what is planned...) and maybe also some hints how to get started with gPTP protocol for INET 4 - if I got it right, the message dispatcher should allow me to grab a message before it will be enqueued so that I can simulate the high priority of gPTP messages?

Best wishes

PS: I am not sure if an issue is the right place for some discussions, if you would like to discuss the topic in the mailing group (if it is still alive?) - just accept my request :)

**Comment by levy** [comment]
Actually IEEE 802.1AS is already implemented and ported to INET 4.x. The model is supposed to be pushed to the master branch today. I'm not saying that all features are implemented, but the basic functionality is working. There is an example demonstrating the effects of clock synchronization under examples/ethernet/clocksynchronization. This example will have a separate configuration for gPTP.

ps: it would be better to discuss this on the mailing list, because then others can also see it and participate.

**Comment by HpLightcorner** [comment]
That are indeed great news!
Looking forward to test the implementation. 

PS: I have a lot of other TSN related (e.g. is there already a proposed way to implement a node which will send time-aware and best effort traffic...) with INET 4.x, but I will ask those questions then in the mailing group.

**Reply by HpLightcorner** [reply]
@rhornig and @levy - thanks in advance for your support! Looking forward to get some input 👍
